File: testgen/ui/components/frontend/js/components/empty_state.js
/**
* @typedef Message
* @type {object}
* @property {string} line1
* @property {string} line2
*
* @typedef Link
* @type {object}
* @property {string} href
* @property {string} label
*
* @typedef Properties
* @type {object}
* @property {string} icon
* @property {string} label
* @property {Message} message
* @property {Link?} link
* @property {any?} button
* @property {string?} class
*/
import van from '../van.min.js';
import { Card } from '../components/card.js';
import { getValue, loadStylesheet } from '../utils.js';
import { Link } from './link.js';

const { i, span, strong } = van.tags;

const EMPTY_STATE_MESSAGE = {
    connection: {
        line1: 'Begin by connecting your database.',
        line2: 'TestGen delivers data quality through data profiling, hygiene review, test generation, and test execution.',
    },
    tableGroup: {
        line1: 'Profile your tables to detect hygiene issues',
        line2: 'Create table groups for your connected databases to run data profiling and hygiene review.',
    },
    profiling: {
        line1: 'Profile your tables to detect hygiene issues',
        line2: 'Run data profiling on your table groups to understand data types, column contents, and data patterns.',
    },
    testSuite: {
        line1: 'Run data validation tests',
        line2: 'Automatically generate tests from data profiling results or write custom tests for your business rules.',
    },
    testExecution: {
        line1: 'Run data validation tests',
        line2: 'Execute tests to assess data quality of your tables.'
    },
    score: {
        line1: 'Track data quality scores',
        line2: 'Create custom scorecards to assess quality of your data assets across different categories.',
    },
    explorer: {
        line1: 'Track data quality scores',
        line2: 'Filter or select columns to assess the quality of your data assets across different categories.',
    },
};

const EmptyState = (/** @type Properties */ props) => {
    loadStylesheet('empty-state', stylesheet);

    return Card({
        class: `tg-empty-state flex-column fx-align-flex-center ${getValue(props.class ?? '')}`,
        content: [
            span({ class: 'tg-empty-state--title mb-5' }, props.label),
            i({class: 'material-symbols-rounded mb-5'}, props.icon),
            strong({ class: 'mb-2' }, props.message.line1),
            span({ class: 'mb-5' }, props.message.line2),
            (
                getValue(props.button) ??
                (
                    getValue(props.link)
                    ? Link({
                        class: 'tg-empty-state--link',
                        right_icon: 'chevron_right',
                        ...(getValue(props.link)),
                    })
                    : ''
                )
            ),
        ],
    });
}

const stylesheet = new CSSStyleSheet();
stylesheet.replace(`
.tg-empty-state {
    margin-top: 80px;
    border: 1px solid var(--border-color);
    padding: 112px 24px !important;
}

.tg-empty-state--title {
    font-size: 24px;
    color: var(--secondary-text-color);
}

.tg-empty-state > i {
    font-size: 60px;
    color: var(--disabled-text-color);
}

.tg-empty-state > .tg-empty-state--link {
    margin: auto;
    border-radius: 4px;
    border: var(--button-stroked-border);
    padding: 8px 8px 8px 16px;
    color: var(--primary-color);
}
`);

export { EMPTY_STATE_MESSAGE, EmptyState };

File: testgen/ui/components/widgets/empty_state.py
import typing
from enum import Enum

import streamlit as st

from testgen.ui.components.widgets.button import button
from testgen.ui.components.widgets.link import link
from testgen.ui.components.widgets.page import css_class, whitespace
from testgen.ui.services.user_session_service import DISABLED_ACTION_TEXT


class EmptyStateMessage(Enum):
    Connection = (
        "Begin by connecting your database.",
        "TestGen delivers data quality through data profiling, hygiene review, test generation, and test execution.",
    )
    TableGroup = (
        "Profile your tables to detect hygiene issues",
        "Create table groups for your connected databases to run data profiling and hygiene review.",
    )
    Profiling = (
        "Profile your tables to detect hygiene issues",
        "Run data profiling on your table groups to understand data types, column contents, and data patterns.",
    )
    TestSuite = (
        "Run data validation tests",
        "Automatically generate tests from data profiling results or write custom tests for your business rules.",
    )
    TestExecution = (
        "Run data validation tests",
        "Execute tests to assess data quality of your tables."
    )


def empty_state(
    label: str,
    icon: str,
    message: EmptyStateMessage,
    action_label: str,
    action_disabled: bool = False,
    link_href: str | None = None,
    link_params: dict | None = None,
    button_onclick: typing.Callable[..., None] | None = None,
    button_icon: str = "add",
) -> None:
    with st.container(border=True):
        css_class("bg-white")
        whitespace(5)
        st.html(f"""
                <div style="text-align: center;">
                    <p style="font-size: 24px; color: var(--secondary-text-color);">{label}</p>
                    <p><i class="material-symbols-rounded" style="font-size: 60px; color: var(--disabled-text-color);">{icon}</i></p>
                    <p><b>{message.value[0]}</b><br>{message.value[1]}</p>
                </div>
                """)
        _, center_column, _ = st.columns([.4, .3, .4])
        with center_column:
            if link_href:
                link(
                    label=action_label,
                    href=link_href,
                    params=link_params or {},
                    right_icon="chevron_right",
                    underline=False,
                    height=40,
                    style=f"""
                        margin: auto;
                        border-radius: 4px;
                        border: var(--button-stroked-border);
                        padding: 8px 8px 8px 16px;
                        color: {"var(--disabled-text-color)" if action_disabled else "var(--primary-color)"};
                    """,
                    disabled=action_disabled,
                    tooltip=DISABLED_ACTION_TEXT if action_disabled else None,
                )
            elif button_onclick:
                button(
                    type_="stroked" if action_disabled else "flat",
                    color="basic" if action_disabled else "primary",
                    label=action_label,
                    icon=button_icon,
                    on_click=button_onclick,
                    style="margin: auto; width: auto;",
                    disabled=action_disabled,
                    tooltip=DISABLED_ACTION_TEXT if action_disabled else None,
                )
        whitespace(5)
